hail/src/main/scala/is/hail/types/physical/PCanonicalNDArray.scala

```diff
@@ -191,43 +326,66 @@ final case class PCanonicalNDArray(elementType: PType, nDims: Int, required: Boo
 
   def setRequired(required: Boolean) = if(required == this.required) this else PCanonicalNDArray(elementType, nDims, required)
 
-  def unstagedStoreAtAddress(addr: Long, region: Region, srcPType: PType, srcAddress: Long, deepCopy: Boolean): Unit =
-    this.fundamentalType.unstagedStoreAtAddress(addr, region, srcPType.fundamentalType, srcAddress, deepCopy)
+  def unstagedStoreAtAddress(addr: Long, region: Region, srcPType: PType, srcAddress: Long, deepCopy: Boolean): Unit = {
+    val srcND = srcPType.asInstanceOf[PCanonicalNDArray]
+
+    if (deepCopy) {
+      region.trackNDArray(srcAddress)
+    }
+    Region.storeAddress(addr, copyFromAddress(region, srcND, srcAddress, deepCopy))
+  }
 
   def sType: SNDArrayPointer = SNDArrayPointer(this)
 
   def loadCheapPCode(cb: EmitCodeBuilder, addr: Code[Long]): PCode = new SNDArrayPointerCode(sType, addr)
 
   def store(cb: EmitCodeBuilder, region: Value[Region], value: SCode, deepCopy: Boolean): Code[Long] = {
     value.st match {
-      case SNDArrayPointer(t) if t.equalModuloRequired(this) =>
-          representation.store(cb, region, representation.loadCheapPCode(cb, value.asInstanceOf[SNDArrayPointerCode].a), deepCopy)
+      case SNDArrayPointer(t) if t.equalModuloRequired(this)  =>
+        val storedAddress = cb.newLocal[Long]("pcanonical_ndarray_store", value.asInstanceOf[SNDArrayPointerCode].a)
+        if (deepCopy) {
+          cb.append(region.trackNDArray(storedAddress))
+        }
+        storedAddress
+      case SNDArrayPointer(t) =>
+        val oldND = value.asNDArray.memoize(cb, "pcanonical_ndarray_store_old")
+        val shape = oldND.shapes(cb)
+        val newStrides = makeColumnMajorStrides(shape, region, cb)
+        val (targetDataFirstElementAddr, finish) = this.constructDataFunction(shape, newStrides, cb, region)
+
+        val currentOffset = cb.newLocal[Long]("pcanonical_ndarray_store_offset", targetDataFirstElementAddr)
+        SNDArray.forEachIndex(cb, shape, "PCanonicalNDArray_store") { (cb, currentIndices) =>
+          val oldElement = oldND.loadElement(currentIndices, cb)
+          elementType.storeAtAddress(cb, currentOffset, region, oldElement, true)
+          cb.assign(currentOffset, currentOffset + elementType.byteSize)
+        }
+
+        finish(cb).a
+      case _ =>
```

Comment:
you can just leave this case out, and when we add something out we'll get a better error (match error with the thing that failed to match)